dsmiddleware/memcache/memcache.go

```diff
@@ -0,0 +1,149 @@
+package memcache
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"github.com/bradfitz/gomemcache/memcache"
+	"go.mercari.io/datastore"
+	"go.mercari.io/datastore/dsmiddleware/storagecache"
+	"time"
+)
+
+var _ storagecache.Storage = &cacheHandler{}
+var _ datastore.Middleware = &cacheHandler{}
+
+const defaultExpiration = 15 * time.Minute
+
+// New memcache middleware creates & returns.
+func New(memcacheAddr string, opts ...CacheOption) interface {
+	datastore.Middleware
+	storagecache.Storage
+} {
+	memcacheClient := memcache.New(memcacheAddr)
+
+	ch := &cacheHandler{
+		client:           memcacheClient,
+		stOpts:         &storagecache.Options{},
+		expireDuration: defaultExpiration,
+	}
+
+	for _, opt := range opts {
+		opt.Apply(ch)
+	}
+
+	s := storagecache.New(ch, ch.stOpts)
+	ch.Middleware = s
+
+	if ch.logf == nil {
+		ch.logf = func(ctx context.Context, format string, args ...interface{}) {}
+	}
+	if ch.cacheKey == nil {
+		ch.cacheKey = func(key datastore.Key) string {
+			return "mercari:memcache:" + key.Encode()
+		}
+	}
+
+	return ch
+}
+
+type cacheHandler struct {
+	datastore.Middleware
+	stOpts *storagecache.Options
+
+	client         *memcache.Client
+	expireDuration time.Duration
+	logf           func(ctx context.Context, format string, args ...interface{})
+	cacheKey       func(key datastore.Key) string
+}
+
+// A CacheOption is an cache option for a memcache middleware.
+type CacheOption interface {
+	Apply(*cacheHandler)
+}
+
+func (ch *cacheHandler) SetMulti(ctx context.Context, cis []*storagecache.CacheItem) error {
+
+	ch.logf(ctx, "dsmiddleware/memcache.SetMulti: incoming len=%d", len(cis))
+
+	for _, ci := range cis {
+		if ci.Key.Incomplete() {
+			panic("incomplete key incoming")
```

Comment:
doesn't support incomplete key?

Reply:
This is how they implement the `Incomplete()` func https://github.com/kailuo-qb/datastore/blob/master/clouddatastore/key.go#L127. I think they do a panic check here because without the ID and name, GobEncode will fail.
It does support incompletekey, because by the time it starts putting the incompletekey into cache, the key should  have the name/key that's filled by datastore. I tested it already.